### PR TITLE
Add shouldComponentUpdate to ContextualMenuItems to ensure items are only updated when needed.

### PR DESCRIPTION
**Pull request by ddlbrena**
Add shouldComponentUpdate to ContextualMenuItems to ensure items are only updated when needed.

This is specifically needed for cases when menus are being persisted.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/8360)

### CODE REVIEW

**Comment by KevinTCoughlin** [comment]
> This is specifically needed for cases when menus are being persisted.

Should we _at least_ add unit test coverage verifying these changes against the above mentioned scenarios?

We'll need to ensure that we don't regress perf here cc: @joschect

**Comment by ddlbrena** [comment]
I will look into adding a test to ensure the component is updated when needed and vice versa. At the moment, the menu item does not know if its parent menu is persisted or not. When menus are persisted this has better perf, else this comparison cost is added. I can go and measure the extra cost that is added by measuring the demo app.

---
In reply to: [474058462](https://github.com/OfficeDev/office-ui-fabric-react/pull/8360#issuecomment-474058462) [](ancestors = 474058462)

**Comment by KevinTCoughlin** [comment]
> When menus are persisted this has better perf, else this comparison cost is added. I can go and measure the extra cost that is added by measuring the demo app.

That would be great! Its the additional cost for the default scenarios here that I am most concerned about cc: @joschect @dzearing.

**Comment by ddlbrena** [comment]
As a side note, ComboBoxOptionWrapper already does this (for a different purpose) and the calls take around 20 micro seconds on average

---
In reply to: [474128368](https://github.com/OfficeDev/office-ui-fabric-react/pull/8360#issuecomment-474128368) [](ancestors = 474128368)

**Comment by joschect** [comment]
@KevinTCoughlin, @ddlbrena the contextualmenu items, and content of the callout aren't actually rendered when it's hidden, so this is effectively only reducing the amount of times that the hidden contextualmenu would be rendered, I think at least. This could enable you to have more hidden menus without slowing down the pages perf overall. It also has a potential side effect of preventing rerenders when the consumer expects rerenders because in the past they might have been able to get away with just updating a property of an object. We may want to warn people, although that's an edge case.

I don't believe there are any scenarios where this could make perf worse, but it can't hurt to do a quick double check. Definitely adding a test makes sense.

@dzearing Should we make a PureBaseComponent or something like that which extends React.PureComponent?

**Comment by KevinTCoughlin** [comment]
> @dzearing Should we make a PureBaseComponent or something like that which extends React.PureComponent?

I think we're trying to get rid of extended some `BaseComponent` altogether in favor of composition and functional helpers.

**Comment by dzearing** [comment]
Yes what @KevinTCoughlin mentioned; we need to remove references to BaseComponent.

**Comment by ddlbrena** [comment]
So I been trying to write a test that verifies that the component updates when the props change and it does not update then the props object changes but all of the values are the same compared to the previous one. I been looking for a way to know when the component renders or componentDidUpdate is called. So far I have not been able to get this to work, any ideas?

**Comment by ddlbrena** [comment]
Ok, I think I got it, will send an update soon.

**Comment by ddlbrena** [comment]
Added tests for just one of the sub-components. Please let me know if that is enough. I could add to all of them but did not think is needed.

File: packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenuItemWrapper/ContextualMenuButton.test.tsx [hunk 1]
```diff
@@ -46,6 +46,50 @@ describe('ContextualMenuButton', () => {
       expect(onClickMock).toHaveBeenCalledTimes(1);
       expect(onClickMock).toHaveBeenCalledWith(menuItem, expect.objectContaining(mockEvent));
     });
+
+    it('does not update when props values do not change', () => {
+      let reactCalls: number = 0;
+      jest.spyOn(ContextualMenuButton.prototype, 'render').mockImplementation(() => {
+        reactCalls += 1;
+        return null;
+      });
```
Comment (KevinTCoughlin) [inline packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenuItemWrapper/ContextualMenuButton.test.tsx]: @ddlbrena would you mind changing these mocks you've added to be `jest` mocks which can be asserted against using `toHaveBeenCalledTimes`, see https://jestjs.io/docs/en/expect#tohavebeencalledtimesnumber.

Will remove the need for an explicit counter. We use them in `List` tests for example: https://github.com/OfficeDev/office-ui-fabric-react/blob/2948f3d6aba40e03e5fafd5de06459597eccd9af/packages/office-ui-fabric-react/src/components/List/List.test.tsx#L125

Comment (KevinTCoughlin) [reply packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenuItemWrapper/ContextualMenuButton.test.tsx]: I'll make the change this evening if you're unable to get to it.

Comment (ddlbrena) [reply packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenuItemWrapper/ContextualMenuButton.test.tsx]: I can do it, I will send an update in a bit.

Comment (ddlbrena) [reply packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenuItemWrapper/ContextualMenuButton.test.tsx]: Done.

Comment (KevinTCoughlin) [reply packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenuItemWrapper/ContextualMenuButton.test.tsx]: Awesome! I pushed up a slight revision that simplifies the test code, see https://github.com/OfficeDev/office-ui-fabric-react/pull/8360/commits/842f3dfb3dc2643e43c50693cd052159739cf5f9.

**Comment by KevinTCoughlin** [comment]
Hopefully @joschect can take a look in the morning. LGTM
